### PR TITLE
Tdd calendar meeting

### PR DESCRIPTION
**Pull request by anthonyqzhu**
Week 5 - implement the meeting query feature to find possible meeting times

### CODE REVIEW

**Comment by yjennywang** [comment]
did you forget to add the files for meeting finder tests?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,80 @@
...
+    Collection<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return possibleTimes;
+    }
+
+    possibleTimes.add(TimeRange.WHOLE_DAY);
+
+    if (request.getAttendees().size() == 0) {
+        return possibleTimes;
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:44]: Should you instead throw an error here?

**Comment by anthonyqzhu** [comment]
> did you forget to add the files for meeting finder tests?

The first tests were already included in FindMeetingQueryTest.java. The next step is to add more tests which will be in the next PR

**Comment by yjennywang** [comment]
> > did you forget to add the files for meeting finder tests?
> 
> The first tests were already included in FindMeetingQueryTest.java. The next step is to add more tests which will be in the next PR

I think you might have forgotten to commit that file. I'm not seeing it in this PR.

Can you git status to see if it needs to be commited? Or is it in another PR?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,80 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:32]: It might be useful to add documentation of what exactly this function does. In particular, how are the function parameters used.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,80 @@
...
+  private void removeInsufficientSlots(Collection<TimeRange> possibleTimes, long requestDuration) {
+      Collection<TimeRange> rangesToRemove = new ArrayList<TimeRange>();
+      for (TimeRange range : possibleTimes) {
+          if (range.duration() < requestDuration) {
+              rangesToRemove.add(range);
+          }
+      }
+
+      possibleTimes.removeAll(rangesToRemove);
   }
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Its a bit hard to follow whats going on here. Could you explain?

**Review by sidharthp — CHANGES_REQUESTED** [review]

**Comment by anthonyqzhu** [comment]
> > > did you forget to add the files for meeting finder tests?
> > 
> > 
> > The first tests were already included in FindMeetingQueryTest.java. The next step is to add more tests which will be in the next PR
> 
> I think you might have forgotten to commit that file. I'm not seeing it in this PR.
> 
> Can you git status to see if it needs to be commited? Or is it in another PR?

It was already there when I cloned the repo and I didn't change it so it was not included in the changes. The code is in the week 5 project src folder under tests. Should I still add it to the PR?

**Comment by yjennywang** [comment]
> > > > did you forget to add the files for meeting finder tests?
> > > 
> > > 
> > > The first tests were already included in FindMeetingQueryTest.java. The next step is to add more tests which will be in the next PR
> > 
> > 
> > I think you might have forgotten to commit that file. I'm not seeing it in this PR.
> > Can you git status to see if it needs to be commited? Or is it in another PR?
> 
> It was already there when I cloned the repo and I didn't change it so it was not included in the changes. The code is in the week 5 project src folder under tests. Should I still add it to the PR?

Ah, my mistake! I thought you were supposed to write the test. :)

Thank you for clarifying!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:44]: I think one of the tests implied that the return value should be an empty collection

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:32]: done in recent commit

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I tried to explain with the new comments - are there any functions that are confusing in particular?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,93 @@
...
+    }
+
+    // Start with the whole day available
+    possibleTimes.add(TimeRange.WHOLE_DAY);
+
+    if (request.getAttendees().size() == 0) {
+        return possibleTimes;
+    }
+
+    // Only update times for events that have attendees from the meeting request
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It doesn't really update times for existing events, so can you clarify the comment?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,93 @@
...
+
+  /** 
+   * Update possible meeting times by excluding time ranges of an event
+   */
+  private void updatePossibleTimes(Collection<TimeRange> possibleTimes, TimeRange eventTime, long requestDuration) {
+    Collection<TimeRange> rangesToModify = new ArrayList<TimeRange>();
+
+    // Check for time ranges that contain the event in question
+    for (TimeRange timeSlot : possibleTimes) {
+        if (timeSlot.contains(eventTime.start()) || timeSlot.contains(eventTime.end())) {
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Could there be a case where timeSlow contains neither `start` nor `end` but still overlaps? For instance, if `timeSlow = 1pm - 2pm` and `eventTime = 1.15pm - 1.45pm`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,93 @@
...
+                updatePossibleTimes(possibleTimes, event.getWhen(), request.getDuration());
+            }
+        }
+    }
+
+    return possibleTimes;
+  }
+
+  /** 
+   * Update possible meeting times by excluding time ranges of an event
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: One thing that would be useful to do here and on the javadoc of `splitTimeRanges` is explain what is happening with an example. For instance, `If possibleTimes={[12am,1am],[3am, 4am]}, eventTime=[3.30-4.30am], requestDuration=1hr -----> return {[12am,1am]}.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,93 @@
...
+    for (Event event : events) {
+        for (String attendee : request.getAttendees()) {
+            if (event.getAttendees().contains(attendee)) {
+                updatePossibleTimes(possibleTimes, event.getWhen(), request.getDuration());
+            }
+        }
+    }
+
+    return possibleTimes;
+  }
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: What is the time-complexity of this approach? Assume that there are N events and M attendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes I think the case would be if the eventTime completely contained the timeSlot time, for example eventTime = 1:00-2:00 but timeSlot = 1:15-1:45 - I think I have addressed this case in the newest commit

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I guess this would be O(NM) because it is searching for M attendees for each event, and the lookup is constant time for a HashSet of attendees. I wasn't sure if there was a better way of doing this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -75,6 +78,10 @@ private void updatePossibleTimes(Collection<TimeRange> possibleTimes, TimeRange
 
   /** 
    * For ranges that overlap the event, remove the invalid parts from the range
+   *
+   * For example, if eventTime = [530-600] and timeSlot = [545-615], adds [600-615]
+   * to possible times and remove timeSlot. If eventTime = [545-615] and timeSlot = [530-600],
+   * then timeSlot would be removed and [530-545] would be added.
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The comment doesn't match the function arguments.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (sidharthp) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think it takes O(N) time to both add and remove an element from an array list, in which case your algorithm might actually be O(N^4). I think its fine to leave it like this for now unless the task is to have an optimized algorithm.

Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: would you suggest using a different data structure in order to reduce the complexity?

Comment (sidharthp) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Using a Set would definitely reduce the time to add and remove from O(N) to O(log N), so you can consider doing that.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -58,56 +58,36 @@
    * Update possible meeting times by excluding time ranges of an event
    *
    * For example, if possibleTimes={[12am,1am],[3am, 4am]}, eventTime=[3.30-4.30am], requestDuration=60min,
-   * then possibleTimes={[12am,1am]} after the function executes.
+   * then {[12am,1am]} is returned after the function executes.
    */
-  private void updatePossibleTimes(Collection<TimeRange> possibleTimes, TimeRange eventTime, long requestDuration) {
-    Collection<TimeRange> rangesToModify = new ArrayList<TimeRange>();
+  private Collection<TimeRange> updatePossibleTimes(Collection<TimeRange> possibleTimes, TimeRange eventTime, long requestDuration) {
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Rename `updatePossibleTimes` to `getNewPossibleTimes`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -58,56 +58,36 @@
...
+    Collection<TimeRange> newTimes = new ArrayList<TimeRange>();
 
     // Check for time ranges that contain the event in question
     for (TimeRange timeSlot : possibleTimes) {
         if (timeSlot.contains(eventTime.start()) || timeSlot.contains(eventTime.end()) || eventTime.contains(timeSlot)) {
```
Comment (sidharthp) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Can you instead use `TimeRange.overlaps`?

**Review by sidharthp — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (yjennywang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Adding to what Sidharth said -- TreeSets also ensure that your time ranges are sorted, which can help with all the range overlap lookups you're doing.

Right now, you're doing O(N) lookups to find the right place to split. Once this is sorted, though, you can essentially do binary searches to find overlapping time ranges (possible time ranges are always non-overlapping, iiuc, and in a strictly increasing order)

Possibly TreeSet.subSet could get you all the overlapping time ranges (you'd essentially have to do some comparator trickery, and the potential for off by one errors abounds). I assume it's O(logn) and binary searching.

Comment (yjennywang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ^ was very much not meant as a blocking comment, but just something to think about. ;)

Another possibility might be to see if there is preprocessing you can do to make the problem less complex. You aren't required to necessarily work with the data structures and orderedness you're given in the parameters.

For instance, if you had two sorted lists/sets of timeranges, possibleTimes and timeRangesToRemove, would it be faster to remove the time ranges then (in the same way that removeAll() can be faster than individual removes)?
